lib/resty/upstream/healthcheck.lua

```diff
@@ -289,85 +294,66 @@ local function check_peer(ctx, id, peer, is_backup)
+        for _, peer in pairs(peers) do
+            check_peer(ctx, peer.id, peer, is_backup)
         end
     else
         local threads
         local nthr
 
         if n <= concur then
-            nthr = n - 1
+            nthr = n
             threads = new_tab(nthr, 0)
-            for i = 1, nthr do
+
+            for k, peer in pairs(peers) do
+                if next(peers, _) == nil then
+                    -- use the current "light thread" to run the last task
+                    if debug_mode then
+                        debug("check ", is_backup and "backup" or "primary", " peer ",
+                            peer.id)
+                    end
+                    check_peer(ctx, peer.id, peer, is_backup)
+                    break
+                end
 
                 if debug_mode then
                     debug("spawn a thread checking ",
-                          is_backup and "backup" or "primary", " peer ", i - 1)
+                          is_backup and "backup" or "primary", " peer ", peer.id)
                 end
 
-                threads[i] = spawn(check_peer, ctx, i - 1, peers[i], is_backup)
-            end
-            -- use the current "light thread" to run the last task
-            if debug_mode then
-                debug("check ", is_backup and "backup" or "primary", " peer ",
-                      n - 1)
+                threads[k] = spawn(check_peer, ctx, peer.id, peer, is_backup)
             end
-            check_peer(ctx, n - 1, peers[n], is_backup)
-
         else
             local group_size = ceil(n / concur)
             local nthr = ceil(n / group_size) - 1
+            local count = 0
 
             threads = new_tab(nthr, 0)
-            local from = 1
-            local rest = n
-            for i = 1, nthr do
-                local to
-                if rest >= group_size then
-                    rest = rest - group_size
-                    to = from + group_size - 1
-                else
-                    rest = 0
-                    to = from + rest - 1
-                end
 
-                if debug_mode then
-                    debug("spawn a thread checking ",
-                          is_backup and "backup" or "primary", " peers ",
-                          from - 1, " to ", to - 1)
-                end
-
-                threads[i] = spawn(check_peer_range, ctx, from, to, peers,
-                                   is_backup)
-                from = from + group_size
-                if rest == 0 then
-                    break
+            while n > 0 do
+                for k, peer in pairs(peers) do
+                    if debug_mode then
+                        debug("spawn a thread checking ",
+                            is_backup and "backup" or "primary", " peer ",
+                            peer.id)
+                    end
+                    threads[k] = spawn(check_peer, ctx, peer.id, peer, is_backup)
```

Comment:
ive changed it back to using check_peer_range and tophatted and it looks good. the conditional that was failing in the tophat before was when #peers <= concurrency